### DISCUSSION TITLE
Platform-specific wheels & fallback

### DISCUSSION THREAD

**Discussion by marcperrinoptel** [devpi/devpi#879] [Q&A] [answered]
Hello,

Not sure it's the right place for such questions (mine is probably quite basic), but:

If I upload a platform-specific wheel (Windows) to my devpi, on Linux the behavior should be to fallback to /root/pypi, right?

In my case, when I remove my uploaded Windows wheel, the behavior on Linux is fine (downloading (...)/root/pypi/(...)), but with the wheel present, I get "Could not find a version that satisfies the requirement", i.e. somehow, having uploaded a Windows wheel blocks the correct Linux process...
What can I look at to troubleshoot that?

Thanks!

**Comment by marcperrinoptel** [comment]
NB: it occurs to me that if my index does not have root/pypi as a parent (which unfortunately I cannot check), then "no fallback" is expected...

**Comment by fschulze** [comment, marked as answer]
It is a security feature: https://devpi.net/docs/devpi/devpi/stable/+doc/userman/devpi_indices.html?highlight=mirror_whitelist#modifying-the-mirror-whitelist

Either add the package in question to the whitelist ``devpi index mirror_whitelist+=pkgname``, or create a separate index just for platform specific wheels and set ``mirror_whitelist`` to ``*`` and inherit from that index. Any wheel uploaded to that index will automatically be whitelisted. I do not recommend that on a regular index where you have private packages.

**Reply by marcperrinoptel** [reply]
OK thanks. So if I've created a separate index with `bases=root/pypi`, it's "not enough" and I still have to set its `mirror_whitelist` to `*`?

**Reply by fschulze** [reply]
Yes and your regular index needs the new index as its base, you can replace the default ``root/pypi`` by the name of your wheels index.

**Reply by marcperrinoptel** [reply]
Thanks - to be thorough, let me summarize the situation/resolution on my side:
- we had a devpi setup with a single "main" index (referenced in the `pip.conf/ini` files of the users) having for bases both `root/pypi` & an index called `release` (`type=stage`) for private packages
- I needed to serve some wheels from https://www.lfd.uci.edu/~gohlke/pythonlibs/, so `release` index was not really the place for that; I ended up putting those Windows-specific wheels in a new index that I called `unofficial` i.e. a 3rd base for the "main" index. Most importantly, the new index has the target packages mentioned in its `mirror_whitelist` to avoid the issue I had originally, i.e. Linux failing to fallback to PyPI (NB: `release` has nothing in `mirror_whitelist`)

**Reply by fschulze** [reply]
I would use linear inheritance:
- ``unofficial`` index ``bases=root/pypi`` and ``mirror_whitelist=packages``
- ``release`` index ``bases=unofficial``  ``mirror_whitelist=``
- ``main`` index ``bases=release``

If that doesn't work for you for some reason (you want the ``release`` index without ``unofficial`` as well):
- ``release`` index ``bases=root/pypi``  ``mirror_whitelist=``
- ``unofficial`` index ``bases=root/pypi`` and ``mirror_whitelist=packages``
- ``main`` index ``bases=unofficial,release``

**Reply by marcperrinoptel** [reply]
Linear makes sense to me. It's just that it's a bigger departure from the situation I found where `main` had both `root/pypi` & `release` as bases. Once I'm more confident with devpi, I might change for that ;-)
